File: app/src/main/java/org/autojs/autojs/ui/BaseActivity.java
package org.autojs.autojs.ui;

import static android.content.pm.PackageManager.PERMISSION_DENIED;

import android.content.Context;
import android.content.res.Configuration;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.Bundle;
import android.view.Menu;
import android.view.View;

import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.appcompat.widget.Toolbar;
import androidx.core.content.ContextCompat;

import com.stardust.app.GlobalAppContext;
import com.stardust.theme.ThemeColorManager;
import com.zeugmasolutions.localehelper.LocaleHelper;
import com.zeugmasolutions.localehelper.LocaleHelperActivityDelegate;
import com.zeugmasolutions.localehelper.LocaleHelperActivityDelegateImpl;

import org.autojs.autojs.Pref;
import org.autojs.autojs.R;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/**
 * Created by Stardust on 2017/1/23.
 * Modified by SuperMonster003 as of Feb 18, 2022.
 */

public abstract class BaseActivity extends AppCompatActivity {

    protected static final int PERMISSION_REQUEST_CODE = 11186;
    private boolean mShouldApplyDayNightModeForOptionsMenu = true;
    private final LocaleHelperActivityDelegate localeDelegate = new LocaleHelperActivityDelegateImpl();
    private final LocaleHelper localeHelper = LocaleHelper.INSTANCE;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        localeDelegate.onCreate(this);
    }

    protected void applyDayNightMode() {
        GlobalAppContext.post(() -> {
            if (Pref.isNightModeEnabled()) {
                setNightModeEnabled(Pref.isNightModeEnabled());
            }
        });
    }

    @NonNull
    @Override
    public AppCompatDelegate getDelegate() {
        return localeDelegate.getAppCompatDelegate(super.getDelegate());
    }

    @Override
    protected void attachBaseContext(Context newBase) {
        super.attachBaseContext(localeDelegate.attachBaseContext(newBase));
        if (localeHelper.hasLocaleSelection(this)) {
            localeHelper.setLocale(this, localeHelper.getLocale(this));
        }
    }

    public void setNightModeEnabled(boolean enabled) {
        if (enabled) {
            getDelegate().setLocalNightMode(AppCompatDelegate.MODE_NIGHT_YES);
        } else {
            getDelegate().setLocalNightMode(AppCompatDelegate.MODE_NIGHT_NO);
        }
        if (getDelegate().applyDayNight()) {
            recreate();
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        if ((getWindow().getDecorView().getSystemUiVisibility() & View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN) == 0) {
            ThemeColorManager.addActivityStatusBar(this);
        }

    }

    @Override
    protected void onResume() {
        super.onResume();
        localeDelegate.onResumed(this);
    }

    @Override
    protected void onPause() {
        super.onPause();
        localeDelegate.onPaused();
    }

    @Override
    public Context createConfigurationContext(Configuration overrideConfiguration) {
        Context context = super.createConfigurationContext(overrideConfiguration);
        return localeHelper.onAttach(context);
    }

    @Override
    public Context getApplicationContext() {
        return localeDelegate.getApplicationContext(super.getApplicationContext());
    }

    public void updateLocale(Locale locale) {
        localeDelegate.setLocale(this, locale);
    }

    public void setLocaleFollowSystem() {
        Locale systemLocale = localeHelper.getSystemLocale();
        Locale currentLocale = localeHelper.getLocale(this);

        if (currentLocale != systemLocale) {
            localeDelegate.setLocale(this, systemLocale);
        }
        localeDelegate.clearLocaleSelection(this);
    }

    public <T extends View> T $(int resId) {
        return findViewById(resId);
    }

    protected void checkPermission(String... permissions) {
        String[] requestPermissions = getRequestPermissions(permissions);
        if (requestPermissions.length > 0) {
            requestPermissions(requestPermissions, PERMISSION_REQUEST_CODE);
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    private String[] getRequestPermissions(String[] permissions) {
        List<String> list = new ArrayList<>();
        for (String permission : permissions) {
            if (checkSelfPermission(permission) == PERMISSION_DENIED) {
                list.add(permission);
            }
        }
        return list.toArray(new String[0]);
    }

    public void setToolbarAsBack(String title) {
        setToolbarAsBack(this, R.id.toolbar, title);
    }

    public static void setToolbarAsBack(final AppCompatActivity activity, int id, String title) {
        Toolbar toolbar = activity.findViewById(id);
        toolbar.setTitle(title);
        activity.setSupportActionBar(toolbar);
        if (getSupportActionBar(activity) != null) {
            toolbar.setNavigationOnClickListener(v -> activity.finish());
            getSupportActionBar(activity).setDisplayHomeAsUpEnabled(true);
        }
    }

    private static ActionBar getSupportActionBar(AppCompatActivity activity) {
        return activity.getSupportActionBar();
    }

    @CallSuper
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        if (mShouldApplyDayNightModeForOptionsMenu && Pref.isNightModeEnabled()) {
            for (int i = 0; i < menu.size(); i++) {
                Drawable drawable = menu.getItem(i).getIcon();
                if (drawable != null) {
                    drawable.mutate();
                    drawable.setColorFilter(ContextCompat.getColor(this, R.color.toolbar), PorterDuff.Mode.SRC_ATOP);
                }
            }
            mShouldApplyDayNightModeForOptionsMenu = false;
        }
        return super.onPrepareOptionsMenu(menu);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        return super.onCreateOptionsMenu(menu);
    }
}

File: app/src/main/java/org/autojs/autojs/ui/BaseActivity.kt
@file:Suppress("DEPRECATION")

package org.autojs.autojs.ui

import android.content.Context
import android.os.Build
import android.os.Bundle
import android.view.View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
import android.view.View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
import android.view.View.SYSTEM_UI_FLAG_LAYOUT_STABLE
import android.view.WindowManager
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import org.autojs.autojs.core.accessibility.AccessibilityTool
import org.autojs.autojs.core.pref.Language
import org.autojs.autojs.theme.ThemeChangeNotifier
import org.autojs.autojs.theme.ThemeColorManager
import org.autojs.autojs.util.LocaleUtils
import org.autojs.autojs.util.ViewUtils
import org.autojs.autojs6.R

/**
 * Created by Stardust on Jan 23, 2017.
 * Modified by SuperMonster003 as of Feb 18, 2022.
 */
abstract class BaseActivity : AppCompatActivity() {

    open val handleContentViewFromHorizontalNavigationBarAutomatically = true
    open val handleStatusBarThemeColorAutomatically = true
    open val handleNavigationBarContrastEnforcedAutomatically = true

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        ViewUtils.addWindowFlags(this, WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)

        @Suppress("DEPRECATION")
        ViewUtils.appendSystemUiVisibility(this, SYSTEM_UI_FLAG_LAYOUT_STABLE or SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION)

        if (handleContentViewFromHorizontalNavigationBarAutomatically) {
            ViewUtils.excludeContentViewFromHorizontalNavigationBar(this)
        }

        if (handleNavigationBarContrastEnforcedAutomatically) {
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
                ViewUtils.setNavigationBarBackgroundColor(this, getColor(R.color.black_alpha_44))
            }
        }

        setApplicationLocale(this)

        ThemeChangeNotifier.themeChanged.observe(this) {
            initThemeColors()
            if (handleStatusBarThemeColorAutomatically) {
                ThemeColorManager.setStatusBarBackgroundColor(this)
                setUpStatusBarIconLightByThemeColor()
            }
        }

        // @Dubious by SuperMonster003 on Oct 28, 2024.
        //  ! Is it property to start a11y service automatically here?
        //  ! zh-CN: 无障碍服务自启动放在这里是否合适?
        AccessibilityTool(this).apply { if (!isServiceRunning()) startService(false) }
    }

    private fun setApplicationLocale(context: Context) {
        when {
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU -> {
                LocaleUtils.setApplicationLocale(AppCompatDelegate.getApplicationLocales())
            }
            else -> when (Language.getPrefLanguage().isAuto()) {
                true -> LocaleUtils.setFollowSystem(context)
                else -> LocaleUtils.setApplicationLocale(Language.getPrefLanguageOrNull()?.locale)
            }
        }
    }

    override fun onStart() {
        super.onStart()
        initThemeColors()
        if (handleStatusBarThemeColorAutomatically) {
            ThemeColorManager.addActivityStatusBar(this)
            setUpStatusBarIconLightByThemeColor()
        }
    }

    override fun onResume() {
        super.onResume()
        LocaleUtils.syncPrefWithCurrentStateIfNeeded(this)
    }

    open fun initThemeColors() {
        /* Empty body. */
    }

    fun setToolbarAsBack(titleRes: Int) = ViewUtils.setToolbarAsBack(this, titleRes)

    fun setToolbarAsBack(title: String?) = ViewUtils.setToolbarAsBack(this, title)

    protected fun setUpStatusBarIconLightByNightMode() {
        ViewUtils.setStatusBarIconLight(this, ViewUtils.isNightModeEnabled)
    }

    protected fun setUpStatusBarIconLightByThemeColor() {
        ThemeColorManager.setStatusBarIconLight(this)
    }

}